### ISSUE TITLE
WaterMeterController Support (YS5008-UC and YS5006-UC)

### ISSUE THREAD

**Issue by ssnthedj** [dkerr64/homebridge-yolink#98]
Please add support for Water Meter controller YS5008-UC

https://shop.yosmart.com/pages/ys5008-support

This is a brand new product from Yolink so I understand why it is not supported yet but it would be awesome if it could be added. I think this product will be quite popular at it's price point. Way cheaper than Moen Flo or Phyn Plus.

**Comment by dkerr64** [comment]
Do you have one to test with?  I would need to see the log messages as the plugin discovered the device and what messages are received.

The plugin already supports the original YoLink water valve (I have a YS4909-UC) and if this new one uses the same basic commands then I would expect open/close to work.  But the additional features that measure the flow will not, and I don't think HomeKit would support them anyway.

**Comment by ssnthedj** [comment]
Yes I do have one, I actually got this error message when I added it to the Yolink app and then looked in my homebridge instance and saw the errors about it not being supported.  I've attached the relevant section of the log files with the error.

[homebridge.log.txt](https://github.com/user-attachments/files/16258783/homebridge.log.txt)

**Comment by dkerr64** [comment]
Thanks for the log.  It is reporting itself as a different device so current plugin will not support.  I should be able to add support but I may need you to do a couple or rounds of testing with me.  I'll let you know when I have something for you to try.

Thanks.

**Comment by ssnthedj** [comment]
Gotcha, yes this is a completely new product from Yolink combining their water meter and smart valve into one device. I think it will be a popular product from them given it's competitive price point.  It's half the cost of others like Moen flow. I'm happy to do any testing on my end just let me know once you need. I'm getting the device installed by a plumber next week.

**Comment by dkerr64** [comment]
I just published a beta version to NPM.  Go to the plugin in Homebridge and select "manage version" from the dot-dot-dot menu.  It is v1.5.8-beta.1

I believe it should correctly report status of the valve.  I am not sure whether it will be able to set the valve (open or close it) as I am guessing that the data to send is the same as what I send for the older style valve.  But that may be a bad assumption... as the format of the status data received from this valve is different (per your log file).  Unfortunately YoLink have not published the API for this device at their [developer API document page](http://doc.yosmart.com/docs/overall/intro) yet.  If it doesn't work, then I will have to reach out to them for documentation.

New feature in this device is that it appears to report whether water is flowing or not.  HomeKit supports this, so I added support for that.  Again, I don't have one to test (and you won't be able to test that before you have it installed!.

Please turn on verbose logging and report back with what it shows.  I'm particularly interested in seeing what is logged when you control the device from your phone YoLink app... as that will cause messages to be sent to the plugin reporting status.

Thanks.

**Comment by ssnthedj** [comment]
I installed v1.5.8-beta.1 and the water meter now shows up in HomeKit and does not give a "device not supported" error in homebridge. However, when I try to control the valve it throws an API error. I am attaching the logs for your review. Also, in HomeKit it keeps showing "starting" on the device but that does not change. 

[homebridge-yolink.log.txt](https://github.com/user-attachments/files/16276003/homebridge-yolink.log.txt)
![IMG_9348](https://github.com/user-attachments/assets/b3a10296-32fc-4d6d-b7ff-12705950469c)
![IMG_9350](https://github.com/user-attachments/assets/5736ff67-cc65-4553-8e9b-fb79ea2c97b7)
![IMG_9349](https://github.com/user-attachments/assets/75df4d24-2453-4264-9289-da4daa01426d)

**Comment by dkerr64** [comment]
I'll need to look and see if I can find some documentation on what this device expects.  Also, did you enable verbose logging in the plugin settings?  If not please do that and try restarting again.

The 00201 errors are, unfortunately, fairly common (see my comments on that in the [technical notes](https://github.com/dkerr64/homebridge-yolink?tab=readme-ov-file#technical-notes) section of README.  So not clear whether they are relevant.

Thanks.

**Comment by ssnthedj** [comment]
I had enabled verbose logging for homebridge itself but did not realize I needed to do separately for plugin.  See snippet below of verbose output when trying to close/open the valve.

<img width="1149" alt="Screenshot 2024-07-18 at 12 20 04 AM" src="https://github.com/user-attachments/assets/1b23ad0c-a82a-4a80-83a7-6b5884ac7960">

**Comment by dkerr64** [comment]
The log looks odd because it is missing data I would expect to see after the SENDING: and RECEIVING: statements (will be JSON strings).  What version of Homebridge UI are you using?  It may be clipping that data when you only view logs for the YoLink plugin.  They are showing fine for me with the UI v4.56.4.

I need to see the full YoLink log to check that the device is properly registered... assuming the log is properly reporting the JSON data.  Can you download and attach that please.  **You should redact any appearance of your _userAccessId_ and _secretKey_** (they may appear twice, once when the config file is logged, and again when they are used. Other identifiers are okay to leave in as they have temporary life... but if you want to be super cautious you can redact the _access_token_... it has a life of a couple of hours.

Also, next time you test, please do not try and control the valve from HomeKit / Apple Home.  Instead use the OEM YoLink app to control the valve... that should trigger messages on the log which I need to see.  I am looking for something like this.

```
[7/18/2024, 8:36:15 AM] [YoLink] [verbose] MQTT message: yl-home/d3d51dfc169c4f7aa02b1bf5901a92db/d88b4c01000369a3/report
{
  "event": "Manipulator.setState",
  "time": 1721306175062,
  "msgid": "1721306175062",
  "data": {
    "state": "open",
    "loraInfo": {
      "netId": "010205",
      "signal": -71,
      "gatewayId": "d88b4c1603008c02",
      "gateways": 1
    }
  },
  "deviceId": "d88b4c01000369a3"
}
```

But yours will say _WaterMeterController_ instead of _Manipulator_.

**Comment by dkerr64** [comment]
I am also looking for something like this during plugin initialization...

```
[7/18/2024, 8:27:37 AM] [YoLink] [verbose] [YS4909-UC (d88b4c01000369a3) Water Main Valve] checkDeviceState (refresh after 3600 seconds)
[7/18/2024, 8:27:37 AM] [YoLink] [YS4909-UC (d88b4c01000369a3) Water Main Valve] YoLinkAPI.getDeviceState [lite]
[7/18/2024, 8:27:37 AM] [YoLink] [verbose] YoLinkAPI.getAccessToken
[7/18/2024, 8:27:37 AM] [YoLink] [verbose] SENDING:
{"time":1721305657484,"method":"Manipulator.getState","targetDevice":"d88b4c01000369a3","token":"C346EFC1AEEADD1D9C69CE30815DBD30"}
[7/18/2024, 8:27:39 AM] [YoLink] [verbose] RECEIVED:
{"code":"000000","time":1721305659209,"msgid":1721305659209,"method":"Manipulator.getState","desc":"Success","data":{"state":"open","battery":4,"powerSupply":"battery","delay":{"ch":1,"off":0},"openRemind":0,"version":"0911","time":"2024-06-18T00:27:38.000Z","tz":-4,"loraInfo":{"netId":"010205","signal":-71,"gatewayId":"d88b4c1603008c02","gateways":1}}}
[7/18/2024, 8:27:39 AM] [YoLink] [verbose] [YS4909-UC (d88b4c01000369a3) Water Main Valve] Battery level: 100%
[7/18/2024, 8:27:39 AM] [YoLink] [YS4909-UC (d88b4c01000369a3) Water Main Valve] At 7/18/2024, 8:27:39 AM: Device state updated: Valve (both): open, Battery: 4 [lite]
[7/18/2024, 8:27:39 AM] [YoLink] Device initialized: YS4909-UC (d88b4c01000369a3) Water Main Valve
[7/18/2024, 8:27:39 AM] [YoLink] [verbose] [YS4909-UC (d88b4c01000369a3) Water Main Valve] Set data refresh timer to run every 3598 seconds
```
Thanks.

**Comment by dkerr64** [comment]
I just pushed a new beta.  Please try it and capture logs.

Thanks.

**Comment by ssnthedj** [comment]
Attached is an updated log. With your latest beta I can actually control the valve now so that's progress! I think there may still be some issues with reporting data to homekit though. For instance I see the words "waiting" and "starting" a lot in the homekit UI for the meter. I have not captured logs when controlling the valve from the yolink app but I can certainly do that if you still need me to. 
[homebridge-yolink.log.txt](https://github.com/user-attachments/files/16301215/homebridge-yolink.log.txt)

**Comment by ssnthedj** [comment]
when I initially change the state of the valve in homekit it updates the status, for instance when I open the valve it says "running" as the status in homekit. However if a leave it for a while and come back it just says "starting" or "waiting". If I close the valve it says "off" in homekit. I have not left it closed to see if I come back if it still says that but I might try and see what happens. It is worth nothing this is not just a smart valve but also a water meter so hopefully get can get that to report as well. Of course that can't really be tested until it is installed into my water line (hopefully next week).

**Comment by dkerr64** [comment]
Good, we are making progress.  The log file is very helpful, I can now see all the data fields and I do see at least one change I'll need to make.

I have the original manipulator valve, and I also see "starting" and "waiting" when I control the valve.  I am not sure what causes those but it may have something to do with the _InUse_ characteristic.  This is supposed to indicate that fluid is flowing through the valve.  For the original YoLink valve it did not report whether water was flowing, so I just assume that if the valve is open then it is _InUse_.   However, your valve does include a _waterFlowing_ value, and I now use that to set the _InUse_ characteristic property... which for now is always false.  But once it is installed, that _waterFlowing_ value should change to true (or false) and hopefully that will make the "starting" and "waiting" states change.

**Comment by dkerr64** [comment]
As for anything other than open/close, our options are limited.  The HomeKit accessory documentation does not include water meters (or power meters).  Some 3rd party apps have added support for device types not in the official spec.  Examples are smart plugs which can measure power consumed by the connected device, but it is hard to implement (and I have not done it for the YoLink smart plugs).  Water meters are likely in the same situation.

It might be possible to add a leak detector accessory, that is in the HomeKit spec, and I do support the YoLink leak detectors.  This water meter is reporting a number of possible alarm conditions, one of which is leak.  So that may be possible.

I'm also interested in the temperature field.  Right now it is presumably reporting ambient temperature... but I'm curious if once you have it installed it will report the water temperature.

**Comment by dkerr64** [comment]
I also see quite a lot of `YoLink API error code: 020104 Device is busy, try again later.` errors in the log.  These will likely go away once you settle down with it installed.  See comments in my [readme](https://github.com/dkerr64/homebridge-yolink?tab=readme-ov-file#network-resiliency).

**Comment by dkerr64** [comment]
>However, your valve does include a _waterFlowing_ value, and I now use that to set the _InUse_ characteristic property... which for now is always false. But once it is installed, that _waterFlowing_ value should change to true (or false) and hopefully that will make the "starting" and "waiting" states change.

However, this may not be desired behavior... if you open a valve that controls your whole house, then water may not flow immediately... another faucet would need to be opened.  We will have to test this once you have it installed and may have to revert back to how I have it for the older valve.

**Comment by tothbe** [comment]
@dkerr64 The temperature field will always show the ambient temperature as it measures that for the anti-freezing function and alerting built in to the device. I am also aiming to buy one soon so I made some investigation on its possibilities.

**Comment by dkerr64** [comment]
I just pushed beta 3 which adds support for the temperature sensor.  Note that temperature sensors (that are embedded within another type of device) are not enabled automatically... you have to request it in the configuration either globally for all devices (that provide temperature) or per-device.  See [README](https://github.com/dkerr64/homebridge-yolink?tab=readme-ov-file#configuration-file). 

As I do not have one of these devices I cannot test this, so I'd appreciate some feedback and a verbose log.

**Comment by dkerr64** [comment]
And beta 4 has leak sensor added.  Again I cannot test this so I'd appreciate feedback and verbose log.

**Comment by ssnthedj** [comment]
Would you like me to install beta 3, collect logs and then install beta 4 and do the same or can I just go direct to beta 4? I do have some of the leak sensors too I was planning to use in tandem with the valve.

**Comment by ssnthedj** [comment]
> > However, your valve does include a _waterFlowing_ value, and I now use that to set the _InUse_ characteristic property... which for now is always false. But once it is installed, that _waterFlowing_ value should change to true (or false) and hopefully that will make the "starting" and "waiting" states change.
> 
> However, this may not be desired behavior... if you open a valve that controls your whole house, then water may not flow immediately... another faucet would need to be opened. We will have to test this once you have it installed and may have to revert back to how I have it for the older valve.

I think that's fine actually because then I can easily see if a device (sink, shower etc) is using water in my house.

**Comment by dkerr64** [comment]
> Would you like me to install beta 3, collect logs and then install beta 4 and do the same or can I just go direct to beta 4? I do have some of the leak sensors too I was planning to use in tandem with the valve.

Just go straight to beta 4.

Thanks

**Comment by ssnthedj** [comment]
See updated logs after upgraded to beta 4. I removed anything prior to the beta 4 upgrade. In homekit it still says a status of "waiting" when the valve is open. Maybe that will change when water is flowing though it? Also I do have one of Yolink's stand alone water leak detectors but I don't see that in homekit, I guess you were talking about the leak detection feature of the water meter/valve itself. 

[homebridge-yolink.log.txt](https://github.com/user-attachments/files/16322389/homebridge-yolink.log.txt)

**Comment by ssnthedj** [comment]
When the valve is closed it does report a status of "off" in homekit but when it's open it just says "waiting".

![Valve Closed](https://github.com/user-attachments/assets/9046cb41-f8fe-4116-90e3-fbf587454f7c)
![Room overview screen](https://github.com/user-attachments/assets/8b648dbb-df01-4c5d-b155-eb2b02fa3c08)
![Valve Open](https://github.com/user-attachments/assets/5d63c9ba-14eb-4caa-8d15-a923ac6242cc)

**Comment by ssnthedj** [comment]
Also, here is how the meter/valve shows up in the Yolink app.

![Overview](https://github.com/user-attachments/assets/a9b9ba04-e141-4112-8583-6cdc93c76626)
![Details 1](https://github.com/user-attachments/assets/bdfe1392-f321-4e91-afcc-8283e34570c7)
![Details 2](https://github.com/user-attachments/assets/50c28bf6-650b-4237-b0ee-014a72075471)

**Comment by ssnthedj** [comment]
Oh I just noticed there is a firmware update already. This is a brand new device from Yolink so that does not surprise me. Let me know if you want me to go ahead and update the firmware. That might break some stuff but better to find out now than later.

**Comment by dkerr64** [comment]
> Oh I just noticed there is a firmware update already. This is a brand new device from Yolink so that does not surprise me. Let me know if you want me to go ahead and update the firmware. That might break some stuff but better to find out now than later.

Yes go ahead and update.  It can take some time because the LoRa is very slow.

**Comment by dkerr64** [comment]
> When the valve is closed it does report a status of "off" in homekit but when it's open it just says "waiting".
> 

Yes I think this is expected until you have it installed... the waterFlowing status is always false... which is what would be expected when the valve is closed.  When you turn it on I think it is waiting for the waterFlowing state to change to true... which is not happening until you have it installed.

**Comment by dkerr64** [comment]
> See updated logs after upgraded to beta 4. I removed anything prior to the beta 4 upgrade. In homekit it still says a status of "waiting" when the valve is open. Maybe that will change when water is flowing though it? Also I do have one of Yolink's stand alone water leak detectors but I don't see that in homekit, I guess you were talking about the leak detection feature of the water meter/valve itself.

Your existing leak detectors should appear in Apple Home as "Alarm sensors" which in a room on MacOS version show in the top right.  Click on that and a window opens listing them, like this...
<img width="708" alt="image" src="https://github.com/user-attachments/assets/d7c6a585-13ba-46a3-aa92-69edb62d56de">

And yes, I have added a leak sensor accessory to the water meter itself.  I do not know how that will appear in Apple Home, if not like above then it may be inside the settings for the valve itself... for example, this dehumidifier has three accessories... the dehumidifier itself, a temperature sensor and a humidistat...
<img width="527" alt="image" src="https://github.com/user-attachments/assets/4d5926c9-41b1-4167-bafe-13afbd2867cd">
I suspect that the water temperature sensor and leak detector may appear there.

**Comment by dkerr64** [comment]
Your most recent log looks good.  I'm assuming you did not enable the temperature sensor as I don't see logs related to that.

I think we're at the point of waiting until this is installed to see how it works in real life.

**Comment by dkerr64** [comment]
How does this device determine if there is a leak?  I'm guessing it is something like measuring a really slow flow rate... or an excessively fast flow rate?  Somehow it would be good to test that a leak alert is properly sent to HomeKit.

**Comment by dkerr64** [comment]
Looking at the YoLink app screenshots, I see that the temperature is shown.  I decided that for this device we should always create a temperature accessory service HomeKit rather than require a config settings.  This is enabled in beta 5.

**Comment by ssnthedj** [comment]
Found the my other water leak sensor in the alarms section as well as some other devices from a different brand I thought were not showing up. Turns out I was just not looking in the right place. so thanks for that. As for the leak sensor built into the water meter itself I'm not sure exactly how that detects leaks. I agree I think it makes sense to enable the temperature sensor by default and I am indeed seeing this after updating to the latest beta. I kinda wish it was measuring the water temp but I guess ambient air temp makes more sense to alert if it's getting near freezing mark. I don't know exactly which day yet but I am hoping to get a plumber to install the valve this week. Once that is done I'll send you some updated logs. I really appreciate all the work you're doing on this.

**Comment by ssnthedj** [comment]
Here's how the device shows up now in Homekit.

![IMG_9377](https://github.com/user-attachments/assets/41ce5dde-6c4d-4792-9d8b-ac623132ef1c)
![IMG_9378](https://github.com/user-attachments/assets/b1468605-c82e-4101-8eb2-eec1679cbccf)

**Comment by dkerr64** [comment]
Looking good.  I will wait until after you have it installed and we confirm our theory on the water flowing / in-use indicator before I release it into the mainline.

Thanks.

**Comment by ssnthedj** [comment]
Just had the valve installed today.  Opened it and turned faucets on and off and captured the logs. Unfortunately homekit does not seem to update if water is flowing so I think we are still missing something. Also, not sure if this matters but it appears the temperature is in fact water and not ambient temp. At least it says "water temp" in the Yolink app. 

[homebridge-yolink.log_07-24-2024.txt](https://github.com/user-attachments/files/16366235/homebridge-yolink.log_07-24-2024.txt)

**Comment by dkerr64** [comment]
Is the Apple Home app still showing "waiting" when you open the valve?

I have to decide what to do with the HomeKit "InUse" characteristic.  For the old valve I set it to in-use whenever the value was open, but for this new valve I use the value of "waterFlowing" in the data we get back from it.  But if that data is not coming in promptly then we may not be able to rely on it.

Your log does show "waterFlowing" as true sometimes, false other times.  But I don't know what Apple Home does with the InUse setting.

I just pushed a beta 6... no material differences, I tweaked the code that handles MQTT messages, and now check if the valve reports a fault or not.

**Comment by ssnthedj** [comment]
So I updated to beta-6 and now the status does eventualy update in homekit but it is far from instant. It look about 30 seconds to update in the Yolink app and I was pressing the manual refresh button so I'm not sure how long it would of taken if I didn't do that. I've attached the latest logs as well as a few screenshots. This device is a bit different since it's not just a valve. It's a water meter with a valve. So I do think it makes sense for "water flowing = in use" however maybe that status could be broken out separately? Maybe as a "sensor" and then have the valve status report "open" or "closed"? I will say I don't really how it always says "starting" in homekit when the valve is open but water is not running. 

[homebridge-yolink.log_07-24-2024_9PM.txt](https://github.com/user-attachments/files/16369029/homebridge-yolink.log_07-24-2024_9PM.txt)

![IMG_9399](https://github.com/user-attachments/assets/af46a592-5871-426e-a07e-08588fbfcbc4)
![IMG_9405](https://github.com/user-attachments/assets/9a47f86e-e427-4864-9c9a-dfde57102714)
![IMG_9404](https://github.com/user-attachments/assets/3f25a3c2-bd76-4ce5-abb6-953dd13bb177)

**Comment by ssnthedj** [comment]
Here's what it looks like in Yolink app. Also it's worth noting even when status is showing as "running" in homekit when you click into the details view of the valve it still says "waiting". 

<img width="520" alt="Screenshot 2024-07-24 at 9 33 39 PM" src="https://github.com/user-attachments/assets/903ed7f2-e68a-494b-a8d7-5f495a6a1d6a">

**Comment by ssnthedj** [comment]
I just did another test this time not opening the Yolink app but just opening homekit and looking at the status of the valve. I turned on a sink and at the 1:30 mark it was still not showing as "running". Then I opened up the Yolink app which showed a status of "water is currently running" then I switched back to the homekit app and the status had updated there. So it seems the Yolink app triggered somekind of status refresh. There is even a manual update refresh button in the valve.

**Comment by ssnthedj** [comment]
The water meter/valve runs on batteries by default but does have the option of connecting a power adapter. I wonder if it would update it's status more often if it was AC powered. It seems like even in the Yolink app the water flowing status is not real-time.

**Comment by dkerr64** [comment]
So, I can explain what is going on, but I'm not sure if there are any changes we can make.  Let me explain how things work...

We can ask the device for current state, and we receive unsolicited messages from the device which report its state.  

When you open up Apple Home it generally requests an update.  In your log it does this only once (timestamp 1:12:49am) where a `WaterMeterController.getState` request is sent by us to the device... and you can see in the log the request sent and the received status.

The reason there is only one instance of this in the log is because I have deliberately throttled the number of times I ask a device for status... HomeKit can be very talkative and YoLink imposed rate limits on their servers, and the LoRa protocol is slow.  So I severely cut back how often I actually ask the device for status, instead relying on the unsolicited updates and providing cached status when HomeKit asks.  In the log HomeKit is actually asking for status 90 times... usually multiple in the space of under a second!

The unsolicited messages come in as MQTT and they take two forms in your log... a `WaterMeterController.Report` event and a `WaterMeterController.getState` event.

Report events are sent regularly by a device to provide updated state, generally they come in regular intervals. There are 8 report events over a couple of hours, all of them report water not flowing.

The getState events appear to be triggered when you use the YoLink app... it is telling all other interested parties (us) that someone (the YoLink app) has requested device status, and provides the status.  There are 4 of these in the log in the space of ~90 seconds.  The first two report water flowing as false, the second two as true.

What I am not seeing in the log is any real time reporting of changes.  Other devices send a StatusChange event, which would come through MQTT as `WaterMeterController.StatusChange` but there are none in the log.  So the device is not reporting in realtime when water starts/stops flowing.

The workaround to this is to add a config option to the plugin settings to change the "refresh after" value. Do this just for this device, so in the config user interface add the device using its device identifier (yours is `d88b4c01000a13fb`) then set the refresh time to something like 30 or 60 seconds, that will result in a lot more requests for status actually going through to the device.

I think I will also to reach out to YoLink and ask.  I have seen them update device firmware in the past to add MQTT messages and it feels like that is what is needed here, polling every 30-60 seconds is really not ideal.

In other words... the plugin is working, it is reporting what it receives from the device.

Thanks.

**Comment by dkerr64** [comment]
@ssnthedj I emailed YoLink yesterday, no reply yet.  In the meantime I have changed the plugin (beta 7) to report _InUse_ based on whether valve is open or closed... the same as I do for the older valves that do not report it.  I also added a config setting to allow users to switch this to use the _waterFlowing_ Status from the valve if desired.

If YoLink update the firmware to accurately report the status change then we can try changing the default.

I've documented this in the Device Notes of the readme.

If this looks good in your testing then I will remove beta tag and publish it in a release update.

**Comment by dkerr64** [comment]
@matrixd2 I see you work on the YoLink home assistant plugin.  I wonder if you could review the discussion in this thread and comment?  Thanks.

**Comment by ssnthedj** [comment]
@dkerr64 thanks for all the updates and testing you've done. I updated to the latest beta and it is reporting water is "on" when valve is open. I think that's better than the using the "water flowing" data is that is so delayed. I appreciate you reaching out to yolink about that. Even in the Yollink app the status of "water flowing" is quite delayed. I've seen it take several minutes to update. Really not ideal at all. Maybe they can make some changes. I do wonder if it was a battery saving measure for it to update less often. I am going to try and order a power adapter for mine because honestly I don't like relying on batteries. Strangely Yolink does not sell one just lists the voltage so I need to find something compatible. 

I did have one thought would it be possible to break out the "water flowing" into a separate sensor in homekit? Right now we have water temp, leak sensor so maybe water flowing could be added there but the main status is still if the value is open/closed?

**Comment by matrixd2** [comment]
@dkerr64 Thank you for the work you are doing now.

Water is considered to be flowing only when the water flow exceeds a certain threshold. So there may be a delay in the water flowing state. 

When the water flowing status changes, the device will not send data to the server, so your mqtt client will not receive the **StatusChange** message, You need to refresh by calling the **getState** API

**Comment by dkerr64** [comment]
>
> I did have one thought would it be possible to break out the "water flowing" into a separate sensor in homekit? Right now we have water temp, leak sensor so maybe water flowing could be added there but the main status is still if the value is open/closed?

Leak and Temperature are easy to break out... as HomeKit has specific services for those.  But the water flowing (or _InUse_) state is built into the HomeKit Valve service and there is no separate service available for this.  So we would have to fake it... call it a _Switch_ or a _Contact Sensor_.  I'm not sure that this buys us anything as unless/until YoLink provide more timely reporting its status.  Unless the value can be relied upon, I think it is misleading to report it at all.

**Comment by dkerr64** [comment]
> @dkerr64 Thank you for the work you are doing now.
> 
> Water is considered to be flowing only when the water flow exceeds a certain threshold. So there may be a delay in the water flowing state.
> 
> When the water flowing status changes, the device will not send data to the server, so your mqtt client will not receive the **StatusChange** message, You need to refresh by calling the **getState** API

@matrixd2 Thank you for confirming what we have observed.  This is unfortunate, hopefully YoLink will update the firmware to provide a StatusChange MQTT message.  Regularly sending _getState_ requests is really not the right way to do this.  With rate limits on YoLink cloud servers and the generally low capacity and slow speed of the LoRa network polling should be a last resort.  That said, I can do it and have documented it as a work-around for now. 

I did send an email to R&D Team (address posted at bottom of the [API page](http://doc.yosmart.com/docs/overall/intro)) brining this to their attention, but no reply yet.

Also, can you confirm how a leak or freeze is notified?  Will I receive a MQTT for those, if so what will the event be?  I have seen **_Alert_** events for other devices, is that what the WaterMeterController will send too... that is what I have assumed and coded to for now.

Thanks.

**Comment by dkerr64** [comment]
@miztahsparklez opened #100 for Yolink YS5006-UC which I closed as duplicate of this issue.  Looking at the JSON status provided for it, there are two notable differences... unlike the YS5008-UC this device does not report `waterFlowing` or `temperature` values.

Current plugin beta should work for the YS5006-UC, but I will have to put in some special handling to not add a temperature accessory to HomeKit.

The JSON also shows `"battery": 4, "powerSupply": "PowerLine",` so we now know what is reported when the valve has permanent power connection... but I don't think there is anything to pass back to HomeKit for this.

**Comment by dkerr64** [comment]
@ssnthedj and @miztahsparklez please try beta.9 that I just published.  I simplified the code a bit and only add temperature sensor if it exists (to properly support the YS5006-UC).

Thanks

**Comment by miztahsparklez** [comment]
Cool, I was able to see a "Running" status with the YS5006.

I was able to dig up some old documentation: https://forum.iobroker.net/assets/uploads/files/1634848447889-apple-spezifikation-homekit.pdf

It seems that the required characteristics are : Active, In Use, and Valve Type.

I'd assume Active is the on/off of the valve.  In Use is if water is flowing (does yolink return a boolean value? I guess you solved this by the "waterflowing" value above.) , and valve type would be genericValve (although, I wonder if irrigation or one of the other valve types would allow for different options).

I know for my LG washer/dryer via homebridge plugin can also show time left, which is likely using Set Duration or Remaining Duration.  However, I wonder if any of these could show how long the water has been running.  Could we use any of the overrun values or possibly show meter information?  I am betting the "waiting..." we are seeing is this time value.

**Comment by dkerr64** [comment]
@miztahsparklez Yes you will see "running" status because I set the InUse characteristic based on the valve being open or closed... we're assuming that water is running when it is open.  The YS-5006 does NOT send a separate `waterFlowing` status (based on the data you provided in #100 ) whereas the TS-5008 does.  But as you can see in the above discussion we have determined that we cannot rely on the `waterFlowing` status to be accurately reported.  I'm documenting it in the readme [here](https://github.com/dkerr64/homebridge-yolink/tree/water-valve?tab=readme-ov-file#water-meter-controller).

I am familiar with the documentation you linked, I used it extensively when creating this plugin.  The duration information is for irrigation valve types, and we do not use that type in this plugin.  Also, I do not know what the `overrunXxx` status values are meant to report.

**Comment by dkerr64** [comment]
Just a note that YoLink have now published some [API documentation](http://doc.yosmart.com/docs/yolinkapi/WaterMeterController) for this device..  So when I get a chance I will review it to see if any changes need to be made to this plugin.

Otherwise, if folks monitoring this thread are happy... I'll switch from beta and release a mainline update.

**Comment by miztahsparklez** [comment]
Sounds good!  I think if you communicate with the expectation of what people will see, I think it's a worthwhile improvement from what it was before and gets the basic point of on/off and in use.

The overrun values correspond to "maximum continuous water usage", "maximum continuous water usage duration", and likely the leak detection schedule.  Usage is measured in volume.  Duration is a time limit.  Schedule is usage over time.  Any of these can be set to trigger the valve to turn off within the Yolink app.

My thought was if you used the water usage duration, there would hopefully be some sort of countdown to 0, or maybe a count up for how long the water has been on (and maybe this is your indicator for determining in use if non zero value).

**Comment by dkerr64** [comment]
Support released in version 1.6.0

**Comment by ssnthedj** [comment]
@dkerr64 thanks for all your heard work on this! I've been out of town and thus have not been able to test beta 10 yet with my device. I will be back Sunday afternoon and should be able to update and test that evening.
